Migrating to scalapb 0.6.6 and sbt-protoc 0.99.12

lokkju opened the pull request:
I went a little bit further then just that, and also refactored the plugin to use the new AutoPlugin architecture, and sbt-scripted for plugin testing.

Since there are some major changes, I'd love if others would like to test and review this extensively.

The README has been updated with build/test/usage details.  It's *much* easier now.

I've disabled all the publish settings, since I don't have ability to test; you'll have to handle that part.

thesamet commented:
Thanks for working on this. Few initial comments:
1. The code generator should remain a library rather than an sbt plugin. The reason is that keeping it as a library allows people who don't use SBT to use it. The SBT tests you added are very useful though - thanks for adding them!
2. No need to comment out the release code.

lokkju commented:
1) how about converting it to a library, a library-codegen, and a plugin?  When using from SBT, having it as a plugin is *immensely* useful, and avoids common issues with configuring it for use.  Having the codegen library would still allow easy use outside sbt.  scalapb is itself more complicated to use than it needs to be - I'd be happy to do the same thing there (make it so it is usable with just an addSbtPlugin line, in addition to as a library)

2) I needed to comment it out for my own testing, I can fix that.

thesamet commented:
Many users are customizing the way the generator works by passing various
options to scalapb.gen, or modifying the directory where files are
generated. The current plug-in implementation won't let them to do so.
Making the plug-in mirror all scalapb options would just be duplicating
functionality. Can you explain why the plug-in is immensely useful? It only
adds 3 lines of hardcoded settings.

On Fri, Oct 20, 2017 at 8:49 PM Loki Coyote <notifications@github.com>
wrote:

>
>    1.
>
>    how about converting it to a library, a library-codegen, and a plugin?
>    When using from SBT, having it as a plugin is *immensely* useful, and
>    avoids common issues with configuring it for use. Having the codegen
>    library would still allow easy use outside sbt. scalapb is itself more
>    complicated to use than it needs to be - I'd be happy to do the same thing
>    there (make it so it is usable with just an addSbtPlugin line, in addition
>    to as a library)
>    2.
>
>    I needed to comment it out for my own testing, I can fix that.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/scalapb/sparksql-scalapb/pull/6#issuecomment-338361509>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AASwrZFfy9sCMLl5wJkdVHRY6Qof9QVdks5suWnfgaJpZM4QAQfR>
> .
>
-- 
-Nadav

lokkju commented:
The plugin adds the requires library dependencies and other required plugins (sbt-scalac) at the correct versions required for compatibility.  I find this useful to get someone started; often times, the docs are quite out of data.

It also adds *default* settings.  These settings can be overridden in the project configuration, just as if it didn't set them at all.  For any expert user, who needs to make customizations, this still lets them.

lokkju commented:
Side note, I almost have a SparkSQL Datasource codegen complete for this as well.  Will allow spark.read.format("com.example.proto.demo.Person").load("filename") for standard delimited protobuf data files.

thesamet commented:
As a maintainer of this project, I'd strongly prefer consistency in how
generators are published and configured. Allowing Sbt plugins for each
generator to override users' PB.targets is a recipe for confusion if there
was more than one Sbt plug-in that did so. Happy to accept the rest of the
updates in this PR.

On Sat, Oct 21, 2017 at 12:23 AM Loki Coyote <notifications@github.com>
wrote:

> The plugin adds the requires library dependencies and other required
> plugins (sbt-scalac) at the correct versions required for compatibility. I
> find this useful to get someone started; often times, the docs are quite
> out of data.
>
> It also adds *default* settings. These settings can be overridden in the
> project configuration, just as if it didn't set them at all. For any expert
> user, who needs to make customizations, this still lets them.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/scalapb/sparksql-scalapb/pull/6#issuecomment-338370937>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AASwrVR4gTH6F72lVxuwCFW5lBNdef1wks5suZwDgaJpZM4QAQfR>
> .
>
-- 
-Nadav

lokkju commented:
Hmm.  Good point; how about each sbt-plugin *adds* to the list by default?  I think I could make so that if the user explicitly set the list of generators, that would take precedence; but if there were no explicit definitions, it would add them.

lokkju commented:
especially if you add multiple code generators, I'm trying to avoid having to customize your list.  This is essentially what plugins like release/publish ones do; default a list of steps, but let you override it.

thesamet commented:
I think it's not worth it. The logic would need to deal with deduping
generators, and we will need each Sbt plug-in to do it in a consistent way.
I dont think this is solving a real pain point for novice users, and it has
the potential to complicate configuration for them down the road. Let's
pass on the Sbt plug-in for now.

On Sat, Oct 21, 2017 at 12:48 AM Loki Coyote <notifications@github.com>
wrote:

> Hmm. Good point; how about each sbt-plugin *adds* to the list by default?
> I think I could make so that if the user explicitly set the list of
> generators, that would take precedence; but if there were no explicit
> definitions, it would add them.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/scalapb/sparksql-scalapb/pull/6#issuecomment-338373134>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AASwre0V-SldH2zFnQeEt26gXZ7xgSXjks5suaHrgaJpZM4QAQfR>
> .
>
-- 
-Nadav

lokkju commented:
I have removed the sbt plugin features, and added the ability to read/write from delimited protobuf files (as generated by the java and python libraries) to/from spark sql dataframes.

I have left alone the Protobuf/Parquet code.

I have upgraded the sparksql version number to 2.2.  Unfortunately, there were breaking interface changes between spark sql 2.1 and 2.2 in the OutputWriter interface, so this breaks compatibility with pre-2.2 spark sql.

lokkju commented:
side note, if we *DO* want to support all 2.x spark versions, @cloud-fan has suggested a method that the spark-avvo project is using: https://github.com/apache/spark/pull/16479#issuecomment-339792885

thesamet commented:
Sorry that this has been standing open for so long. It looks however there is no demand for this functionality so I am considering to close this PR. In the latest version of sparksql-scalapb it is easy to convert a proto RDD to dataframe and use that representation to save to parquet - so I wonder if this is still needed.

thesamet commented:
Looks like this is no longer relevant. Please file a new issue if there's anything that is still missing.
